oauth2_http/java/com/google/auth/oauth2/S2A.java

```diff
@@ -0,0 +1,133 @@
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.auth.oauth2;
+
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.json.JsonObjectParser;
+import com.google.api.client.util.GenericData;
+import com.google.auth.http.HttpTransportFactory;
+import com.google.common.collect.Iterables;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Utilities to fetch the S2A (Secure Session Agent) address from the mTLS configuration.
+ *
+ * <p>mTLS configuration is queried from the MDS MTLS Autoconfiguration endpoint.
+ */
+@ThreadSafe
+public final class S2A {
+  public static final String S2A_CONFIG_ENDPOINT_POSTFIX =
+      "/computeMetadata/v1/instance/platform-security/auto-mtls-configuration";
+
+  public static final String METADATA_FLAVOR = "Metadata-Flavor";
+  public static final String GOOGLE = "Google";
+  private static final String PARSE_ERROR_S2A = "Error parsing S2A Config from MDS JSON response.";
+  private static final String MDS_MTLS_ENDPOINT = ComputeEngineCredentials.getMetadataServerUrl() + S2A_CONFIG_ENDPOINT_POSTFIX;
+
+  private S2AConfig config;
+
+  private transient HttpTransportFactory transportFactory;
+
+  public S2A(Optional<HttpTransportFactory> transportFactory) {
+    this.transportFactory = transportFactory.get();
+    this.config = getS2AConfigFromMDS();
+  }
+
+  /** @return the mTLS S2A Address from the mTLS config. */
+  public String getMtlsS2AAddress() {
+    return config.getMtlsAddress();
+  }
+
+  /** @return the plaintext S2A Address from the mTLS config. */
+  public String getPlaintextS2AAddress() {
+    return config.getPlaintextAddress();
+  }
+
+  /**
+   * Queries the MDS mTLS Autoconfiguration endpoint and returns the {@link S2AConfig}.
+   *
+   * <p>Returns {@link S2AConfig} with empty addresses on error.
+   *
+   * @return the {@link S2AConfig}.
```

Comment:
Can you add a more descriptive javadocs that details when empty addresses occur and what users are expected to do with it? What if only plaintextAddress is empty or mtsAddress is empty? Is that different when both are empty?

Reply:
Done in https://github.com/googleapis/google-auth-library-java/pull/1400/commits/fb577a1b6f3ba7b34f6fa0318309deb7efb76db8

Also, what if only plaintextAddress is empty or mtsAddress is empty? Is that different when both are empty?
> The behavior we will add to the Client libraries (specifically, we are adding logic to decide to create channel using S2A in GAX layer) is: 
> try to fetch mtlsS2AAdress first (call `getMtlsS2AAddress()`. if that is empty or if there is a failure when trying to load the mTLS-to-MDS credentials, then fallback to fetch plaintextS2AAddress (call `getPlaintextS2AAddress()`) . If the plaintextS2AAdress is also empty, that means there is some error when trying to talk the MDS, or S2A is not available in that environment, thus we should not try and use S2A.

Reply:
```        
plaintextS2AAddress =
            OAuth2Utils.validateString(responseData, "plaintext_address", PARSE_ERROR_S2A);
        mtlsS2AAddress = OAuth2Utils.validateString(responseData, "mtls_address", PARSE_ERROR_S2A);
      } catch (IOException e) {
...
      }
```

If the response data is something like:
```
{
   plaintext_address: null,
   mtls_address: "localhost",
}
```

then I believe an IOException would be thrown and mtls_address wouldn't be set even though it's valid. Just want confirm the behavior.

Reply:
Yes this behavior could happen where an IOException is thrown and mtls_address is not set even though it is able to be parsed (I don't think it's too likely since we own the MDS mTLS autoconfiguration endpoint, and the json format is tested). 

In https://github.com/googleapis/google-auth-library-java/pull/1400/commits/0e6f5cef7a289b2a875a4788a4eca2d17545c944 I've adjusted the code so that it catches and ignores any exception thrown when parsing the addresses. So in this case, if for whatever reason, parsing the plaintext_address address fails, we still try and parse the mtls_address address. I chatted with @xmenxk, and we agree that in this case we shouldn't really retry since if we get a successful response from the MDS, it is unlikely to change if we query it again.